mars/services/scheduling/supervisor/autoscale.py

```diff
@@ -0,0 +1,269 @@
+
+    @abstractmethod
+    async def stop(self):
+        """Stop auto scale"""
+
+
+class PendingTaskBacklogStrategy(AbstractScaleStrategy):
+    _task: Optional[asyncio.Task]
+
+    def __init__(self, autoscale_conf: Dict[str, Any], autoscaler):
+        self._autoscaler = autoscaler
+        self._scheduler_check_interval = autoscale_conf.get('scheduler_check_interval', 1)
+        self._scheduler_backlog_timeout = autoscale_conf.get('scheduler_backlog_timeout', 20)
+        self._sustained_scheduler_backlog_timeout = autoscale_conf.get(
+            'sustained_scheduler_backlog_timeout', self._scheduler_backlog_timeout)
+        # Make worker_idle_timeout greater than scheduler_backlog_timeout to
+        # avoid cluster fluctuate back and forth。
+        self._worker_idle_timeout = autoscale_conf.get(
+            'worker_idle_timeout', 2 * self._scheduler_backlog_timeout)
+        self._min_workers = autoscale_conf.get('min_workers', 1)
+        assert self._min_workers >= 1, 'Mars need at least 1 worker.'
+        self._max_workers = autoscale_conf.get('max_workers', 100)
+        self._task = None
+
+    @classmethod
+    async def create(cls, autoscale_conf: Dict[str, Any], autoscaler):
+        return cls(autoscale_conf, autoscaler)
+
+    async def start(self):
+        self._task = asyncio.create_task(self._run())
+
+    async def _run(self):
+        try:
+            if self._autoscaler.get_dynamic_worker_nums() < self._min_workers:
+                logger.info(f'Start to request %s initial workers.', self._min_workers)
+                initial_worker_addresses = await asyncio.gather(*[
+                    self._autoscaler.request_worker() for _ in range(
+                        self._min_workers - self._autoscaler.get_dynamic_worker_nums())])
+                logger.info(f'Finished requesting %s initial workers %s',
+                            len(initial_worker_addresses), initial_worker_addresses)
+            while True:
+                await asyncio.sleep(self._scheduler_check_interval)
+                await self._run_round()
+        except Exception as e:  # pragma: no cover
+            logger.exception('Exception occurred when try to auto scale')
+            self._task.cancel()
+            raise e
+
+    async def _run_round(self):
+        queueing_refs = list(self._autoscaler.queueing_refs.values())
+        if any([await queueing_ref.all_bands_busy() for queueing_ref in queueing_refs]):
+            await self._scale_out(queueing_refs)
+        else:
+            await self._scale_in()
+
+    async def _scale_out(self, queueing_refs):
+        logger.info("Try to scale out, current dynamic workers %s", self._autoscaler.get_dynamic_worker_nums())
+        start_time = time.time()
+        await self._autoscaler.request_worker()
+        await asyncio.sleep(self._scheduler_backlog_timeout)
+        rnd = 1
+        while any([await queueing_ref.all_bands_busy() for queueing_ref in queueing_refs]):
```

Comment:
Do we need to check if some tasks exist in the ready queue to decide if we need to scale out?

Reply:
What do you mean ready queue? Every tasks in `queue_ref._band_queues` are not ready and need extra slot for execution. If tasks already accepted by `global_slot_manager` . There is already a slot for the task. Requesting extra slot won't help. If the task execute slowly. We may need scale vertically which is another topic.